services/cd-service/pkg/mapper/environments_config.go

```diff
@@ -140,9 +140,47 @@ func MapEnvironmentsToGroups(envs map[string]config.EnvironmentConfig) []*api.En
 		}
 	}
 	calculateEnvironmentPriorities(tmpEnvs) // note that `tmpEnvs` were copied by reference - otherwise this function would have no effect on `result`
+
+	{
+		var downstreamDepth uint32 = 0
+		for _, group := range result {
+			downstreamDepth = max(downstreamDepth, group.DistanceToUpstream)
+		}
+
+		for _, group := range result {
+			group.Priority = calculateGroupPriority(group.DistanceToUpstream, downstreamDepth)
+		}
+	}
+
 	return result
 }
 
+func calculateGroupPriority(distanceToUpstream, downstreamDepth uint32) api.Priority {
+	lookup := [][]api.Priority {
+		[]api.Priority {api.Priority_YOLO},
+		[]api.Priority {api.Priority_UPSTREAM, api.Priority_PROD},
+		[]api.Priority {api.Priority_UPSTREAM, api.Priority_PRE_PROD, api.Priority_PROD},
+		[]api.Priority {api.Priority_UPSTREAM, api.Priority_PRE_PROD, api.Priority_CANARY, api.Priority_PROD},
+		[]api.Priority {api.Priority_UPSTREAM, api.Priority_OTHER, api.Priority_PRE_PROD, api.Priority_CANARY, api.Priority_PROD},
+	}
+	if downstreamDepth > 4 {
```

Comment:
```suggestion
	if downstreamDepth > len(lookup) - 1 {
```
Please do not use static values, think about the extensibility of the code 😄 The lookup already provides de maximum chain of possible unique priorities

Reply:
That's a good point.